### PR TITLE
doc: clarify docstring of lmul!/rmul!

### LINKED ISSUE #614: Documentation for lmul! and rmul! may cause confusion
Currently, the [documentation](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/index.html#LinearAlgebra.lmul!) for `lmul!` and `rmul!` both have a section that states:

> lmul!(A, B)
> Calculate the matrix-matrix product AB, overwriting B, and return the result.
> 
> rmul!(A, B)
> Calculate the matrix-matrix product AB, overwriting A, and return the result.

which gives the impression of a in-place version of `A .= A*B` or `B .= A*B`, for arbitrary matrices, but that's not what these functions do, and the example does list the `LinearAlgebra.UpperTriangular` usage; additionally

```
a = rand(2,2)
b = rand(2,2)
lmul!(a,b)
```
gives
> ERROR: MethodError: no method matching lmul!(::Array{Float64,2}, ::Array{Float64,2})

Maybe it's better to state that explicitly in the docstring?

### LINKED ISSUE #31207: Documentation for lmul! and rmul! may cause confusion
Currently, the [documentation](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/index.html#LinearAlgebra.lmul!) for `lmul!` and `rmul!` both have a section that states:

> lmul!(A, B)
> Calculate the matrix-matrix product AB, overwriting B, and return the result.
> 
> rmul!(A, B)
> Calculate the matrix-matrix product AB, overwriting A, and return the result.

which gives the impression of a in-place version of `A .= A*B` or `B .= A*B`, for arbitrary matrices, but that's not what these functions do, and the example does list the `LinearAlgebra.UpperTriangular` usage; additionally

```
a = rand(2,2)
b = rand(2,2)
lmul!(a,b)
```
gives
> ERROR: MethodError: no method matching lmul!(::Array{Float64,2}, ::Array{Float64,2})

Maybe it's better to state that explicitly in the docstring?

### PR DESCRIPTION
**Pull request by dkarrasch**
Closes JuliaLang/LinearAlgebra.jl#614.

As usual, first attempt to get started. There is a ton of different methods for `lmul!`/`rmul!`, but I hope that the hint to the triangular structure and the type is adequate and (nearly) exhaustive.

Shall we have a second example in the `jldoctest`, say with a `Diagonal` matrix? Or some less trivial example?

### CODE REVIEW

File: stdlib/LinearAlgebra/src/matmul.jl [hunk 1]
```diff
@@ -178,6 +178,8 @@ mul!(C::AbstractMatrix, A::AbstractVecOrMat, B::AbstractVecOrMat) = generic_matm
     rmul!(A, B)
 
 Calculate the matrix-matrix product ``AB``, overwriting `A`, and return the result.
+Here, `B` must have triangular/diagonal structure, which is additionally encoded
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/matmul.jl]: Maybe use a formulation that is more open since e.g. also orthogonal matrices can be applied in-place.

**Review by andreasnoack — COMMENTED** [review]
Thanks. Just a minor suggestion. Adding more doctests is almost always a good idea as well.

File: stdlib/LinearAlgebra/src/matmul.jl (hunk 1 above)
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/matmul.jl]: Maybe something like "Must have special structure, which is additionally encoded ... and which allows for in-place multiplication."? I realized in https://github.com/JuliaLang/julia/pull/31285#issuecomment-471977636 that we may extend the in-place multiplication to a couple of factorizations involving triangular and orthogonal matrices, basically building on what we have already. Do you think that's of interest to be included in `LinearAlgebra`?

**Comment by dkarrasch** [comment]
I think this is ready for review/merge.

**Pull request merged**